### PR TITLE
add dsn, domain and evm support

### PR DESCRIPTION
**Pull request by DaMandal0rian**
This PR add support for multiple domains such as core-evm and introduces DSN node key support.
- change provider for infrastructure
- add capability to launch a domain i.e evm-core or payments
- add dsn bootstrap node

### CODE REVIEW

**Review by vedhavyas — COMMENTED** [review]
The changes seem to introduce EVM along side Payments. This does not work since node does not support running both domains. The ideal final state we want is to update the current scripts to start a single domain but introduce more confiugurations. Example: we have `payments` built into scripts since there was not other domain available when then. Now we have multiple domains. We need to make those configurable like we did DomainID

File: network-primitives/variables.tf [hunk 1]
```diff
@@ -50,7 +50,26 @@ variable "rpc-node-config" {
     prune               = bool
     node-dsn-port       = number
     enable-domains      = bool
-    domain-id           = number
+  })
+}
+
+variable "domain-node-config" {
+  description = "RPC node deployment config"
```
Comment (DaMandal0rian) [inline network-primitives/variables.tf]: typo to fix

File: resources/gemini-3d/main.tf [hunk 2]
```diff
@@ -16,24 +16,41 @@ module "gemini-3d" {
...
     nodes-per-region    = 0
-    additional-node-ips = var.hetzner_rpc_node_ips
+    additional-node-ips = var.rpc_node_ips
+    docker-org          = "subspace"
+    docker-tag          = "gemini-3d-2023-may-15"
+    domain-prefix       = "eu"
+    reserved-only       = false
+    prune               = false
+    node-dsn-port       = 30434
+    enable-domains      = false
```
Comment (DaMandal0rian) [inline resources/gemini-3d/main.tf]: does `enable-domains` need to be configurable on rpc node too?

File: network-primitives/scripts/create_domain_node_compose_file.sh [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        echo '      "--",'
+        echo "      \"--chain=\$NETWORK_NAME\","
+        echo '      "--validator",'
+        echo '      "--state-pruning", "archive",'
+        echo '      "--blocks-pruning", "archive",'
+        echo '      "--domain-id", "3",'
+        echo '      "--base-path", "/var/subspace/core_evm_domain",'
+        echo '      "--keystore-path", "/var/subspace/keystore",'
+        echo '      "--rpc-cors", "all",'
+        echo '      "--rpc-port", "6933",'
```
Comment (DaMandal0rian) [inline network-primitives/scripts/create_domain_node_compose_file.sh]: fix port to match caddy

File: network-primitives/scripts/create_domain_node_compose_file.sh [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  archival-node:
+    image: ghcr.io/\${NODE_ORG}/node:\${NODE_TAG}
+    volumes:
+      - archival_node_data:/var/subspace:rw
+    restart: unless-stopped
+    ports:
+      - "30333:30333"
+      - "\${NODE_DSN_PORT}:30433"
+    labels:
+      caddy_0: \${DOMAIN_PREFIX}-\${NODE_ID}.\${NETWORK_NAME}.subspace.network
```
Comment (vedhavyas) [inline network-primitives/scripts/create_domain_node_compose_file.sh]: i dont think we need to expose the primary rpc here. It is already exposed under rpc node anyway

File: network-primitives/scripts/create_domain_node_compose_file.sh [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      echo '      "--blocks-pruning", "archive",'
+      echo '      "--base-path", "/var/subspace/system_domain",'
+      echo '      "--keystore-path", "/var/subspace/keystore",'
+      echo '      "--rpc-cors", "all",'
+      echo '      "--rpc-port", "8933",'
+      echo '      "--ws-port", "8944",'
+      echo '      "--no-private-ipv4",'
+      echo '      "--unsafe-ws-external",'
+      echo "      \"--relayer-id=\$RELAYER_SYSTEM_ID\","
+      if [ "${domain_id}" == 1 ]; then
```
Comment (vedhavyas) [inline network-primitives/scripts/create_domain_node_compose_file.sh]: no, this is not the way. 
We made it configurable so we can avoid these if else. We can totally remove these IMO

File: resources/gemini-3d/main.tf (hunk 2 above)
Comment (vedhavyas) [reply resources/gemini-3d/main.tf]: We need to remove all the domain related stuff from RPC including System domain. Seems like we still have system domain running on RPC

**Review by vedhavyas — COMMENTED** [review]
Looks good. There are still domain related infra under RPC. We should remove those as well

File: network-primitives/bootstrap_node_provisioner.tf [hunk 6]
```diff
@@ -156,7 +174,7 @@ resource "null_resource" "start-boostrap-nodes" {
       "systemctl reenable subspace.service",
 
       # set hostname
-      "hostnamectl set-hostname ${var.network-name}-bootstrap-node-${count.index}",
```
Comment (vedhavyas) [inline network-primitives/bootstrap_node_provisioner.tf]: why change the hostname for bootstrap node with domain prefix ?

File: network-primitives/dns.tf [hunk 7]
```diff
@@ -18,10 +18,10 @@ resource "cloudflare_record" "system-domain-rpc" {
   type    = "A"
 }
 
-resource "cloudflare_record" "core-payments-domain-rpc" {
+resource "cloudflare_record" "core-domain-rpc" {
   count   = length(local.rpc_node_ip_v4)
```
Comment (vedhavyas) [inline network-primitives/dns.tf]: this is still referring to rpc instead of domain ip_v4

File: network-primitives/farmer_node_provisioner.tf [hunk 8]
```diff
@@ -156,7 +156,7 @@ resource "null_resource" "start-farmer-nodes" {
       "systemctl reenable subspace.service",
 
       # set hostname
-      "hostnamectl set-hostname ${var.network-name}-farmer-node-${count.index}",
```
Comment (vedhavyas) [inline network-primitives/farmer_node_provisioner.tf]: not sure this change is required here

File: network-primitives/full_node_provisioner.tf [hunk 9]
```diff
@@ -156,7 +156,7 @@ resource "null_resource" "start-full-nodes" {
       "systemctl reenable subspace.service",
 
       # set hostname
-      "hostnamectl set-hostname ${var.network-name}-full-node-${count.index}",
```
Comment (vedhavyas) [inline network-primitives/full_node_provisioner.tf]: Same as above. Why change the hostnames of primary with domain specific details ?

File: network-primitives/rpc_node_provisioner.tf [hunk 10]
```diff
@@ -168,7 +168,7 @@ resource "null_resource" "start-rpc-nodes" {
       "systemctl reenable subspace.service",
 
       # set hostname
-      "hostnamectl set-hostname ${var.network-name}-rpc-node-${count.index}",
```
Comment (vedhavyas) [inline network-primitives/rpc_node_provisioner.tf]: Same as above

**Review by vedhavyas — COMMENTED** [review]
Some minor nits. 
We are good to merge after. Thanks!

File: network-primitives/rpc_node_provisioner.tf (hunk 10 above)
Comment (DaMandal0rian) [reply network-primitives/rpc_node_provisioner.tf]: I am trying to make the hostnames as close as possible to dns and also machine names in provider dashboard.

File: network-primitives/bootstrap_node_provisioner.tf (hunk 6 above)
Comment (DaMandal0rian) [reply network-primitives/bootstrap_node_provisioner.tf]: I am trying to make the hostnames as close as possible to dns and also machine names in provider dashboard.

Comment (vedhavyas) [reply network-primitives/bootstrap_node_provisioner.tf]: Sure but this not represent any closer TBH. Domain prefix is used for RPC 
Are you trying to set the where node is located on hetzner?

**Review by vedhavyas — APPROVED** [review]

**Pull request merged**